### PR TITLE
Add new Layout/PrivateConstant cop

### PR DESCRIPTION
**Pull request by geshwho**
This cop checks that there are no constants defined in the private or
protected blocks of a class. This does not actually make the constant
private and it will still be publicly accessible. The cop will ignore
constants that are explicitly marked with `private_constant`.

Addresses a closed issue raised in rubocop-hq/ruby-style-guide#536.

-----------------
Hi team! I recently came across some code that inspired this cop while reviewing some code. Placing constants in the private block can be misleading because it can lead the person reading/writing it to think that the constant is not publicly accessible when this is in fact not true. The issue linked above suggested using consistent classes (e.g. `Layout/ClassStructure`). This works well, but is a much broader and stricter rule than this proposition.

Let me know what you think, and if this is something that you'd consider adding to the project. Thanks!

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
I think this logic belongs to the existing `Lint/UselessAccessModifier` cop. Currently it checks for class methods, but this problem you outline is pretty much the same.

**Comment by mtamhankar1** [comment]
> I think this logic belongs to the existing `Lint/UselessAccessModifier` cop. Currently it checks for class methods, but this problem you outline is pretty much the same.

:+1: I did not know about this cop. I can take a stab at extending it to support constants in addition to class methods.

File: lib/rubocop/cop/layout/private_constant.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                add_offense(child)
+              end
+            end
+          end
+        end
+
+        def explicit_private_declaration(node)
+          constant_name = node.children[1]
+          node.parent.children.any? do |n|
+            NodePattern.new(
```
Comment (rrosenblum) [inline lib/rubocop/cop/layout/private_constant.rb:83]: This `NodePattern` is being redefined for each iteration of the array. I think you could improve the performance of this by moving the `NodePattern` declaration outside of the this loop.

```ruby
constant_name = node.children[1]
pattern = NodePattern.new("(send nil? :private_constant (:sym :#{constant_name}))")
node.parent.children.any? do |n|
  pattern.match(n)
end
```

File: lib/rubocop/cop/layout/private_constant.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Layout
+      # Placing constants in the private section does not make them private,
```
Comment (rrosenblum) [inline lib/rubocop/cop/layout/private_constant.rb:6]:
```suggestion
      # Placing constants in the `private` section does not make them private,
```

File: lib/rubocop/cop/layout/private_constant.rb (hunk 1 above)
Comment (marcandre) [reply lib/rubocop/cop/layout/private_constant.rb:83]: It should actually be compiled *once*, using `%`...

**Comment by bbatsov** [comment]
I'm closing this PR due to no recent activity. Feel free to re-open it if you ever come back to it.

**Pull request closed without merging**